File: drone_control/src/camera_viewer.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
import numpy as np
import threading
from collections import OrderedDict


class CameraViewer(Node):
    # Label rendering constants
    _LABEL_CHAR_WIDTH = 14
    _LABEL_PADDING = 10
    _LABEL_HEIGHT = 36
    _LABEL_FONT_SCALE = 0.9
    _LABEL_THICKNESS = 2

    # How often (seconds) to scan for new camera topics
    _DISCOVERY_PERIOD = 2.0

    def __init__(self):
        super().__init__('camera_viewer')

        # Parameters
        self.declare_parameter('window_width', 1600)
        self.declare_parameter('window_height', 900)

        self.window_width = self.get_parameter('window_width').value
        self.window_height = self.get_parameter('window_height').value

        self.bridge = CvBridge()
        self.images = OrderedDict()
        self.lock = threading.Lock()

        # Track already-subscribed topics so we don't double-subscribe
        self._subscribed_topics: set = set()
        self._subs_lock = threading.Lock()
        self._subscriptions: list = []

        # Initial discovery + periodic re-scan timer
        self._discover_cameras()
        self._discovery_timer = self.create_timer(self._DISCOVERY_PERIOD, self._discover_cameras)

        # Display loop runs in a separate thread so ROS spin is not blocked
        self._running = True
        self._display_thread = threading.Thread(target=self._display_loop, daemon=True)
        self._display_thread.start()

    def _is_camera_topic(self, topic: str, types: list) -> bool:
        """Return True if the topic is an active UAV camera image topic."""
        if 'sensor_msgs/msg/Image' not in types:
            return False
        parts = topic.split('/')
        # Expect at least /uav<N>/<camera>/.../image_raw
        if len(parts) < 3:
            return False
        if not parts[1].startswith('uav'):
            return False
        if 'image_raw' not in parts:
            return False
        return True

    def _discover_cameras(self) -> None:
        """Scan all published topics and subscribe to any new camera topics."""
        topic_list = self.get_topic_names_and_types()
        for topic, types in topic_list:
            if not self._is_camera_topic(topic, types):
                continue
            with self._subs_lock:
                if topic in self._subscribed_topics:
                    continue
                self._subscribed_topics.add(topic)

            sub = self.create_subscription(
                Image,
                topic,
                lambda msg, t=topic: self._image_callback(msg, t),
                10,
            )
            self._subscriptions.append(sub)
            self.get_logger().info(f'Subscribed to camera topic: {topic}')

    def _image_callback(self, msg: Image, topic: str) -> None:
        try:
            cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding='bgr8')
            with self.lock:
                self.images[topic] = cv_image
        except Exception as e:
            self.get_logger().warn(f'Failed to convert image from {topic}: {e}')

    def _display_loop(self) -> None:
        cv2.namedWindow('Drone Cameras', cv2.WINDOW_NORMAL)
        cv2.resizeWindow('Drone Cameras', self.window_width, self.window_height)

        while self._running and rclpy.ok():
            with self.lock:
                images_snapshot = dict(self.images)

            canvas = self._build_canvas(images_snapshot)
            cv2.imshow('Drone Cameras', canvas)

            key = cv2.waitKey(30) & 0xFF
            if key == 27:  # ESC
                self.get_logger().info('ESC pressed – shutting down camera_viewer')
                self._running = False
                rclpy.shutdown()
                break

        cv2.destroyAllWindows()

    def _build_canvas(self, images_dict: dict) -> np.ndarray:
        canvas = np.zeros((self.window_height, self.window_width, 3), dtype=np.uint8)

        if not images_dict:
            text = 'Waiting for camera streams...'
            (text_w, text_h), _ = cv2.getTextSize(
                text, cv2.FONT_HERSHEY_SIMPLEX, 1.2, 2
            )
            x = (self.window_width - text_w) // 2
            y = (self.window_height + text_h) // 2
            cv2.putText(
                canvas,
                text,
                (x, y),
                cv2.FONT_HERSHEY_SIMPLEX,
                1.2,
                (100, 100, 100),
                2,
            )
            return canvas

        num = len(images_dict)

        # Compute grid dimensions
        if num <= 2:
            cols, rows = num, 1
        elif num <= 4:
            cols, rows = 2, 2
        else:
            cols = 3
            rows = (num + 2) // 3

        cell_w = self.window_width // cols
        cell_h = self.window_height // rows

        for idx, (topic, image) in enumerate(images_dict.items()):
            if image is None:
                continue

            row = idx // cols
            col = idx % cols

            resized = cv2.resize(image, (cell_w, cell_h))

            y0 = row * cell_h
            x0 = col * cell_w
            canvas[y0:y0 + cell_h, x0:x0 + cell_w] = resized

            # Label: use last segment of the topic path as the camera name
            label = topic.split('/')[-2] if topic.count('/') >= 2 else topic
            bg_w = len(label) * self._LABEL_CHAR_WIDTH + self._LABEL_PADDING
            cv2.rectangle(canvas, (x0, y0), (x0 + bg_w, y0 + self._LABEL_HEIGHT), (0, 0, 0), -1)
            cv2.putText(
                canvas,
                label,
                (x0 + 6, y0 + 26),
                cv2.FONT_HERSHEY_SIMPLEX,
                self._LABEL_FONT_SCALE,
                (0, 255, 0),
                self._LABEL_THICKNESS,
            )

        return canvas

    def destroy_node(self) -> None:
        self._running = False
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = CameraViewer()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: drone_control/src/camera_viewer.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/qos.hpp>

#include <cv_bridge/cv_bridge.hpp>
#include <sensor_msgs/msg/image.hpp>

#include <opencv2/opencv.hpp>

#include <map>
#include <mutex>
#include <string>
#include <vector>

class CameraViewer : public rclcpp::Node {
private:
  std::map<std::string, cv::Mat> images_rgb_;
  std::mutex image_mutex_;

  std::vector<rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr> subs_;

  int window_width_{1600};
  int window_height_{900};

public:
  CameraViewer() : Node("camera_viewer") {
    declare_parameter<int>("window_width", 1600);
    declare_parameter<int>("window_height", 900);
    window_width_ = get_parameter("window_width").as_int();
    window_height_ = get_parameter("window_height").as_int();

    cv::namedWindow("Drone Cameras", cv::WINDOW_NORMAL);
    cv::resizeWindow("Drone Cameras", window_width_, window_height_);

    const std::vector<std::string> topics = {
        "/uav1/rgbd_front/color/image_raw",
        "/uav1/rgbd_down/color/image_raw",
    };

    auto qos = rclcpp::SensorDataQoS();

    for (const auto &topic : topics) {
      auto sub = this->create_subscription<sensor_msgs::msg::Image>(
          topic, qos,
          [this, topic](const sensor_msgs::msg::Image::ConstSharedPtr msg) {
            this->imageCallback(msg, topic);
          });
      subs_.push_back(sub);
      RCLCPP_INFO(get_logger(), "Subscribed to: %s", topic.c_str());
    }
  }

  void spinRender() {
    // WallRate não depende do contexto ROS estar válido (evita crash no Ctrl+C)
    rclcpp::WallRate rate(30);

    while (rclcpp::ok()) {
      rclcpp::spin_some(shared_from_this());

      // snapshot das imagens
      std::map<std::string, cv::Mat> images_copy;
      {
        std::lock_guard<std::mutex> lock(image_mutex_);
        images_copy = images_rgb_;
      }

      cv::Mat canvas_bgr = createCanvasBGR(images_copy);
      cv::imshow("Drone Cameras", canvas_bgr);

      int key = cv::waitKey(1) & 0xFF;
      if (key == 27) {  // ESC
        rclcpp::shutdown();
        break;
      }

      // Se Ctrl+C aconteceu durante o loop, sai sem tentar dormir
      if (!rclcpp::ok()) {
        break;
      }
      rate.sleep();
    }
  }

private:
  void imageCallback(const sensor_msgs::msg::Image::ConstSharedPtr &msg,
                     const std::string &topic) {
    RCLCPP_INFO_THROTTLE(get_logger(), *get_clock(), 2000,
                         "RX %s %ux%u encoding=%s",
                         topic.c_str(), msg->width, msg->height, msg->encoding.c_str());

    try {
      // Guarda como RGB
      auto cv_ptr = cv_bridge::toCvShare(msg, "rgb8");
      const cv::Mat &rgb = cv_ptr->image;

      std::lock_guard<std::mutex> lock(image_mutex_);
      images_rgb_[topic] = rgb.clone();
    } catch (const std::exception &e) {
      RCLCPP_ERROR(get_logger(), "cv_bridge error on %s: %s", topic.c_str(), e.what());
    }
  }

  cv::Mat createCanvasBGR(const std::map<std::string, cv::Mat> &images_rgb) {
    // fundo preto
    cv::Mat canvas(window_height_, window_width_, CV_8UC3, cv::Scalar(0, 0, 0));

    // Queremos sempre 2 “slots” (front e down)
    struct Slot { std::string topic; std::string label; };
    const std::vector<Slot> slots = {
        {"/uav1/rgbd_front/color/image_raw", "rgbd_front"},
        {"/uav1/rgbd_down/color/image_raw",  "rgbd_down"},
    };

    const int cols = 2;
    const int rows = 1;
    const int cell_w = window_width_ / cols;
    const int cell_h = window_height_ / rows;

    for (int i = 0; i < (int)slots.size(); ++i) {
      int col = i % cols;
      int row = i / cols;
      int x0 = col * cell_w;
      int y0 = row * cell_h;

      auto it = images_rgb.find(slots[i].topic);
      if (it == images_rgb.end() || it->second.empty()) {
        cv::putText(canvas, ("waiting: " + slots[i].label),
                    cv::Point(x0 + 20, y0 + 40),
                    cv::FONT_HERSHEY_SIMPLEX, 1.0, cv::Scalar(0, 0, 255), 2);
        continue;
      }

      // resize RGB
      cv::Mat resized_rgb;
      cv::resize(it->second, resized_rgb, cv::Size(cell_w, cell_h));

      // converte RGB->BGR para mostrar
      cv::Mat resized_bgr;
      cv::cvtColor(resized_rgb, resized_bgr, cv::COLOR_RGB2BGR);

      resized_bgr.copyTo(canvas(cv::Rect(x0, y0, cell_w, cell_h)));

      cv::putText(canvas, slots[i].label,
                  cv::Point(x0 + 20, y0 + 40),
                  cv::FONT_HERSHEY_SIMPLEX, 1.0, cv::Scalar(0, 255, 0), 2);
    }

    // linha separadora vertical
    cv::line(canvas, cv::Point(cell_w, 0), cv::Point(cell_w, window_height_),
             cv::Scalar(200, 200, 200), 2);

    return canvas;
  }
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<CameraViewer>();
  node->spinRender();
  rclcpp::shutdown();
  return 0;
}